lib/i18n/index.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Return the entered locale or the first locale that the browser provides.
+ * @param  {String} locale
+ * @return {String}
+ */
+export const formatLocale = (locale) => {
+    const firstLanguage =
+        window.navigator.languages && window.navigator.languages.length ? window.navigator.languages[0] : null;
+    return locale || firstLanguage || window.navigator.userLanguage || window.navigator.language;
+};
+
+const upperCaseLocale = (locale = '') => (locale === 'en' ? 'us' : locale).toUpperCase();
+
+/**
+ * We expect it as fr_FR not fr-FR
+ * @param {String} locale
+ * @returns {string}
+ */
+const getTransformedLocale = (locale = '') => {
+    const changedLocale = locale.replace('-', '_');
+    // OS is in French (France) => navigator.language === fr
+    if (changedLocale.length === 2) {
+        return `${changedLocale}_${upperCaseLocale(changedLocale)}`;
+    }
+    return changedLocale;
+};
+
+const getBrowserLocale = () => {
+    // Doesn't work on IE11 ;)
+    try {
+        const queryParams = new window.URL(window.location).searchParams;
+        return getTransformedLocale(formatLocale(queryParams.get('language')));
+    } catch (e) {
+        // Match: xx_XX xx-XX xx 1192
+        const [, locale] = window.location.search.match(/language=(([a-z]{2,}(_|-)[A-Z]{2,})|([a-z]{2,}))/) || [];
+        return getTransformedLocale(formatLocale(locale));
+    }
+};
+
+/**
+ * Return a matched locale if it exists in our supported translations.
+ * Otherwise returns the default translation.
+ * @param {String} browserLocale e.g. en_GB
+ * @returns {String}
+ */
+const getTranslation = (browserLocale = '', translations = []) => {
+    // Check if the full locale exists in the translations.
+    if (translations.includes(browserLocale)) {
+        return browserLocale;
+    }
+
+    // Try again, but only match on the language, not on the locale.
+    const browserLanguage = browserLocale.substr(0, 2);
+    const translationByLanguage = translations.find((lang) => lang.substr(0, 2) === browserLanguage);
+    if (translationByLanguage) {
+        return translationByLanguage;
+    }
+
+    return DEFAULT_TRANSLATION;
+};
+
+export const getLocale = (lang, translations) => {
+    const browser = getBrowserLocale();
+    const locale = getTranslation(lang || browser, translations);
+    const language = locale.substr(0, 2);
+    return { browser, locale, language };
+};
+
+export async function loadLocale({ TRANSLATIONS = [] } = {}, lang) {
+    const { locale, browser, language } = getLocale(lang, TRANSLATIONS);
+
+    CACHE.translationLocale = locale;
+    CACHE.browserLocale = browser;
+    document.documentElement.lang = language;
+
+    // If it's the default translation, we don't need to load the json since it's loaded by default.
+    if (locale === DEFAULT_TRANSLATION) {
+        useLocale(locale);
+        return { locale, browser, language };
+    }
+
+    try {
+        const rep = await fetch(`/settings/i18n/${locale}.json`);
```

Comment:
Can we use dynamic imports so that webpack can resolve the paths (and complain if it doesn't exist) instead of doing raw requests like this?

Perhaps resolving the translation files actually belongs in the application

Also /settings/ won't work for other apps

Reply:
> Can we use dynamic imports so that webpack can resolve the paths (and complain if it doesn't exist) instead of doing raw requests like this?

No need to use webpack to load a static asset. 
Here the only issue is the path

Reply:
The advantages with using webpack is compilation errors rather than runtime errors, and automatic cache busting

Reply:
> The advantages with using webpack is compilation errors  
we load locale by a var, the JSON is created by a cli command. So we won't have issues, cache busting can be done without webpack with less complexity than using it.